server/src/main/java/org/elasticsearch/action/admin/cluster/snapshots/delete/DeleteSnapshotRequest.java

```diff
@@ -47,13 +49,19 @@ public DeleteSnapshotRequest(StreamInput in) throws IOException {
         super(in);
         repository = in.readString();
         snapshots = in.readStringArray();
+        if (in.getTransportVersion().onOrAfter(TransportVersions.DELETE_SNAPSHOTS_ASYNC_ADDED)) {
+            waitForCompletion = in.readBoolean();
+        }
     }
 
     @Override
     public void writeTo(StreamOutput out) throws IOException {
         super.writeTo(out);
         out.writeString(repository);
         out.writeStringArray(snapshots);
+        if (out.getTransportVersion().onOrAfter(TransportVersions.DELETE_SNAPSHOTS_ASYNC_ADDED)) {
+            out.writeBoolean(waitForCompletion);
+        }
```

Comment:
bwc nit: if `request#waitForCompletion == false` but we're sending to an older version that doesn't understand this new field then we shouldn't simply drop it like this. Instead it'd be better to check `ClusterState#getMinTransportVersion` in `TransportDeleteSnapshotAction#doExecute` similarly to `org.elasticsearch.action.admin.cluster.allocation.TransportGetAllocationStatsAction#doExecute`, and fail the request if the cluster hasn't been fully upgraded yet. Then we can at least assert that this case doesn't happen here (but ideally also throw something):

```suggestion
        if (out.getTransportVersion().onOrAfter(TransportVersions.DELETE_SNAPSHOTS_ASYNC_ADDED)) {
            out.writeBoolean(waitForCompletion);
        } else {
            assert waitForCompletion;
        }
```